### PR TITLE
Attribute allocation bug fixes

### PR DESCRIPTION
**Pull request by Pessimistress**
#### Change List
- Fix a bug where `Attribute.allocate` may skip allocation if previously set external buffer or constant value is large enough
- Make BitmapLayer attributes use pre-allocated typed arrays
- Unit tests

### CODE REVIEW

File: modules/core/src/lib/attribute.js [hunk 1]
```diff
@@ -101,10 +100,6 @@ export default class Attribute extends BaseAttribute {
     return needsRedraw;
   }
 
-  getInstanceCount() {
```
Comment (xintongxia) [inline modules/core/src/lib/attribute.js]: This is not used anywhere else?

File: modules/core/src/lib/attribute.js [hunk 2]
```diff
@@ -436,8 +423,7 @@ export default class Attribute extends BaseAttribute {
     const state = this.userData;
 
     // Check that either 'accessor' or 'update' is a valid function
```
Comment (xintongxia) [inline modules/core/src/lib/attribute.js]: comment is not applied

File: modules/core/src/lib/attribute.js [hunk 3]
```diff
@@ -200,15 +192,15 @@ export default class Attribute extends BaseAttribute {
...
-        if (shouldCopy) {
+        if (shouldCopy && oldValue) {
           // Upload the full existing attribute value to the GPU, so that updateBuffer
           // can choose to only update a partial range.
           // TODO - copy old buffer to new buffer on the GPU
           this.buffer.subData(oldValue);
         }
       }
 
-      this.setNeedsUpdate(true, {startRow: instanceCount});
```
Comment (xintongxia) [inline modules/core/src/lib/attribute.js]: Why redraw is not needed?

File: modules/core/src/lib/attribute.js (hunk 1 above)
Comment (Pessimistress) [reply modules/core/src/lib/attribute.js]: No.

File: modules/core/src/lib/attribute.js (hunk 3 above)
Comment (Pessimistress) [reply modules/core/src/lib/attribute.js]: Redraw and update are two different things. `allocate` is only called during update. The line was excessive.

**Pull request merged**